File: business-tests-python/run_all.py
"""
Run every scenario in this folder, show each one's output, and end
with a summary table. Exit code is 0 if every scenario passed, 1 if
any failed.

Usage:

    python run_all.py              # run every tier
    python run_all.py --tier 1     # run only tier 1
    python run_all.py --only 05    # run scripts whose filename contains 05
"""

import argparse
import subprocess
import sys
import time
from pathlib import Path


HERE = Path(__file__).resolve().parent
VENV_PYTHON = HERE / "venv" / "bin" / "python"


def discover(tier, only):
    tier_dirs = sorted(p for p in HERE.glob("tier*") if p.is_dir())
    if tier is not None:
        tier_dirs = [d for d in tier_dirs if d.name == f"tier{tier}"]

    scripts = []
    for td in tier_dirs:
        for f in sorted(td.glob("*.py")):
            if f.name.startswith("_") or f.name.startswith("."):
                continue
            if only and only not in f.name:
                continue
            scripts.append(f)
    return scripts


def main():
    ap = argparse.ArgumentParser(description="Run every Kavach business-tests scenario.")
    ap.add_argument("--tier", type=int, choices=[1, 2, 3], default=None,
                    help="run only scripts in tier<N>/.")
    ap.add_argument("--only", type=str, default=None,
                    help="filter by substring match on filename.")
    args = ap.parse_args()

    if not VENV_PYTHON.exists():
        print(f"[ERROR] venv python not found at {VENV_PYTHON}", file=sys.stderr)
        print("        create it first: python -m venv venv && ./venv/bin/pip install kavach-sdk",
              file=sys.stderr)
        return 1

    scripts = discover(args.tier, args.only)
    if not scripts:
        print("No scenarios matched. Nothing to run.")
        return 0

    bar = "=" * 72
    print(bar)
    print(f"Running {len(scripts)} scenario(s) with {VENV_PYTHON}")
    if args.tier is not None:
        print(f"  tier filter   : tier {args.tier}")
    if args.only:
        print(f"  substring filter: {args.only!r}")
    print(bar)

    results = []
    for script in scripts:
        print()
        print(bar)
        print(f"> {script.relative_to(HERE)}")
        print(bar)
        start = time.monotonic()
        result = subprocess.run([str(VENV_PYTHON), str(script)], cwd=str(HERE))
        elapsed = time.monotonic() - start
        results.append((script, result.returncode == 0, elapsed))

    print()
    print(bar)
    print("Summary")
    print(bar)
    for script, ok, elapsed in results:
        mark = "PASS" if ok else "FAIL"
        rel = str(script.relative_to(HERE))
        print(f"  [{mark}] {rel:<54} {elapsed:>6.2f}s")
    print(bar)

    passed = sum(1 for _, ok, _ in results if ok)
    total = len(results)
    total_time = sum(e for _, _, e in results)
    if passed == total:
        print(f"{passed}/{total} scenarios passed in {total_time:.2f}s")
    else:
        failing = total - passed
        print(f"{passed}/{total} scenarios passed, {failing} FAILED, total {total_time:.2f}s")
    print(bar)

    return 0 if passed == total else 1


if __name__ == "__main__":
    sys.exit(main())

File: business-tests-node/run_all.ts
/**
 * Run every scenario in this folder, show each one's output, and end
 * with a summary table. Exit code is 0 if every scenario passed, 1 if
 * any failed.
 *
 * Usage:
 *
 *   node dist/run_all.js              # run every tier
 *   node dist/run_all.js --tier 1     # run only tier 1
 *   node dist/run_all.js --only 05    # run scripts whose filename contains 05
 */

import { spawnSync } from 'child_process';
import { readdirSync, statSync } from 'fs';
import { join, resolve } from 'path';

const HERE = resolve(__dirname);

function parseArgs(): { tier: number | null; only: string | null } {
  const argv = process.argv.slice(2);
  let tier: number | null = null;
  let only: string | null = null;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i]!;
    if (a === '--tier') {
      const next = argv[i + 1];
      if (next !== undefined) {
        tier = parseInt(next, 10);
        i++;
      }
    } else if (a === '--only') {
      const next = argv[i + 1];
      if (next !== undefined) {
        only = next;
        i++;
      }
    }
  }
  return { tier, only };
}

function discover(tier: number | null, only: string | null): string[] {
  const allEntries = readdirSync(HERE);
  let tierDirs = allEntries
    .filter(name => name.startsWith('tier'))
    .map(name => join(HERE, name))
    .filter(path => {
      try {
        return statSync(path).isDirectory();
      } catch {
        return false;
      }
    })
    .sort();
  if (tier !== null) {
    tierDirs = tierDirs.filter(d => d.endsWith(`tier${tier}`));
  }

  const scripts: string[] = [];
  for (const td of tierDirs) {
    const files = readdirSync(td)
      .filter(f => f.endsWith('.js'))
      .filter(f => !f.startsWith('_') && !f.startsWith('.'))
      .sort();
    for (const f of files) {
      if (only && !f.includes(only)) continue;
      scripts.push(join(td, f));
    }
  }
  return scripts;
}

function rel(path: string): string {
  return path.startsWith(HERE) ? path.slice(HERE.length + 1) : path;
}

function main(): number {
  const { tier, only } = parseArgs();
  const scripts = discover(tier, only);
  if (scripts.length === 0) {
    console.log('No scenarios matched. Nothing to run.');
    return 0;
  }

  const bar = '='.repeat(72);
  console.log(bar);
  console.log(`Running ${scripts.length} scenario(s)`);
  if (tier !== null) console.log(`  tier filter   : tier ${tier}`);
  if (only) console.log(`  substring filter: '${only}'`);
  console.log(bar);

  const results: { script: string; ok: boolean; elapsed: number }[] = [];
  for (const script of scripts) {
    console.log();
    console.log(bar);
    console.log(`> ${rel(script)}`);
    console.log(bar);
    const start = Date.now();
    const r = spawnSync(process.execPath, [script], { stdio: 'inherit', cwd: HERE });
    const elapsed = (Date.now() - start) / 1000;
    results.push({ script, ok: r.status === 0, elapsed });
  }

  console.log();
  console.log(bar);
  console.log('Summary');
  console.log(bar);
  for (const { script, ok, elapsed } of results) {
    const mark = ok ? 'PASS' : 'FAIL';
    const r = rel(script).padEnd(54);
    console.log(`  [${mark}] ${r} ${elapsed.toFixed(2).padStart(6)}s`);
  }
  console.log(bar);

  const passed = results.filter(r => r.ok).length;
  const total = results.length;
  const totalTime = results.reduce((acc, r) => acc + r.elapsed, 0);
  if (passed === total) {
    console.log(`${passed}/${total} scenarios passed in ${totalTime.toFixed(2)}s`);
  } else {
    const failing = total - passed;
    console.log(`${passed}/${total} scenarios passed, ${failing} FAILED, total ${totalTime.toFixed(2)}s`);
  }
  console.log(bar);

  return passed === total ? 0 : 1;
}

process.exit(main());
